### PR TITLE
Add `.from_pretrained` to HookedSAE

### PR DESCRIPTION
**Pull request by ArthurConmy**
# Description

This PR add `.from_pretrained` to load SAEs faster, ahead of a PR with a demo of steering with SAEs.

It adds Connor and Rob's SAEs, as well as https://huggingface.co/nev/gpt2_xl_saes-saex-test/tree/main from Stepan's saex library; https://github.com/neverix/saex

It also updates the loading from HF util to deal with safetensors.
It also updates the HookedSAETransformer notebook.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Screenshots
Please attach before and after screenshots of the change if applicable.

# Checklist:

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have not rewritten tests relating to key interfaces which would affect backward compatibility

### CODE REVIEW

**Review by danbraunai-apollo — COMMENTED** [review]
I like the interface, but I don't like the general setup of only supporting specific SAEs that are specified in main. It invites too much work on TLens developers to try and support the SAEs that people want to use, of which there will be far more than there are models.

I'd suggest putting the onus on others to have their SAEs be in a format that will work with TLens (and making a PR for alterations to Tlens that are needed to support them). And just having from_pretrained take a HF path. This would make it easier for people to use their own SAEs and wouldn't burden TLens as much.

File: transformer_lens/loading_from_pretrained.py [hunk 1]
```diff
@@ -2691,3 +2692,159 @@ def get_basic_config(model_name: str, **kwargs) -> Config:
...
+    sae_alias_map = {}
+    for official_sae_name in OFFICIAL_SAE_NAMES:
+        aliases = SAE_ALIASES.get(official_sae_name, [])
+        for alias in aliases:
+            sae_alias_map[alias] = official_sae_name
+        sae_alias_map[official_sae_name] = official_sae_name
+    return sae_alias_map
+
+
+def convert_ckkissane_sae_cfg(ckkisane_sae_cfg: dict) -> HookedSAEConfig:
```
Comment (ckkissane) [inline transformer_lens/loading_from_pretrained.py:2775]: typo: "ckkisane_sae_cfg" should be "ckkissane_sae_cfg". Same typo in get_pretrained_sae_config

File: transformer_lens/loading_from_pretrained.py [hunk 2]
```diff
@@ -2691,3 +2692,159 @@ def get_basic_config(model_name: str, **kwargs) -> Config:
...
+        raise ValueError(f"Expected fp32 dtype, got {ckkisane_sae_cfg['enc_dtype']}")
+    return HookedSAEConfig(
+        d_sae=ckkisane_sae_cfg["dict_size"],
+        d_in=ckkisane_sae_cfg["act_size"],
+        hook_name=ckkisane_sae_cfg["act_name"],
+        dtype=torch.float32,
+    )
+
+
+def convert_nev_sae_cfg(nev_sae_cfg: dict) -> HookedSAEConfig:
```
Comment (ckkissane) [inline transformer_lens/loading_from_pretrained.py:2786]: Are we sure this uses pre encoder bias? HookedSAE expects this

Comment (ArthurConmy) [reply transformer_lens/loading_from_pretrained.py:2786]: > "apply_b_dec_to_input": true,

From https://huggingface.co/nev/gpt2_xl_saes-saex-test/blob/main/cfg.json

**Comment by neelnanda-io** [comment]
I like Dan's suggestion. I suggest that we allow TL to support specified
SAEs, but to fall back to assuming a default format given a HuggingFace
path. One option is that we make EVERY SAE in the library satisfy this
format, and for legacy SAEs or new SAEs from other parties (eg nnsight
users) we reupload copies to HuggingFace in the right format. This may be
too much effort though.

I am imagining the main use case being people using certain popular public
SAEs rather than users training and hosting their own - maybe this is
incorrect

**Comment by bryce13950** [comment]
Closing this out, since the underlying class is no longer part of TransformerLens

**Pull request closed without merging**